src/modules/commander/state_machine_helper.cpp

```diff
@@ -1102,135 +1102,98 @@ void reset_offboard_loss_globals(actuator_armed_s &armed, const bool old_failsaf
 
+	// User warning
+	switch (battery_warning) {
 	case battery_status_s::BATTERY_WARNING_LOW:
-		mavlink_log_critical(mavlink_log_pub, "Low battery level! Return advised\t");
+		mavlink_log_warning(mavlink_log_pub, "Low %s, return advised\t", battery_level);
 		events::send(events::ID("commander_bat_low"), {events::Log::Warning, events::LogInternal::Info},
-			     "Low battery level! Return advised");
+			     "Low battery level, return advised");
 		break;
 
 	case battery_status_s::BATTERY_WARNING_CRITICAL:
+		mavlink_log_critical(mavlink_log_pub, "Critical %s, return now\t", battery_level);
+		events::send(events::ID("commander_bat_crit"), {events::Log::Critical, events::LogInternal::Info},
+			     "Critical battery level, return now");
+		break;
 
-		static constexpr char battery_critical[] = "Critical battery level!";
-
-		switch (low_battery_action) {
-		case LOW_BAT_ACTION::WARNING:
-			mavlink_log_critical(mavlink_log_pub, "%s, landing advised\t", battery_critical);
-			events::send(events::ID("commander_bat_crit"), {events::Log::Critical, events::LogInternal::Info},
-				     "Critical battery level! Landing advised");
-			break;
+	case battery_status_s::BATTERY_WARNING_EMERGENCY:
+		mavlink_log_emergency(mavlink_log_pub, "Dangerous %s, land now\t", battery_level);
+		events::send(events::ID("commander_bat_emerg"), {events::Log::Emergency, events::LogInternal::Info},
+			     "Dangerous battery level, land now");
+		break;
 
-		case LOW_BAT_ACTION::RETURN:
+	case battery_status_s::BATTERY_WARNING_FAILED:
+		mavlink_log_emergency(mavlink_log_pub, "Battery failure detected, land and check battery\t");
+		events::send(events::ID("commander_bat_failure"), events::Log::Emergency,
+			     "Battery failure detected, land and check battery");
+		break;
 
-		// FALLTHROUGH
-		case LOW_BAT_ACTION::RETURN_OR_LAND:
+	case battery_status_s::BATTERY_WARNING_NONE: break; // no warning
+	}
+}
 
-			if (status_flags.condition_global_position_valid && status_flags.condition_home_position_valid) {
-				if (!(internal_state.main_state == commander_state_s::MAIN_STATE_AUTO_RTL ||
-				      internal_state.main_state == commander_state_s::MAIN_STATE_AUTO_LAND ||
-				      internal_state.main_state == commander_state_s::MAIN_STATE_AUTO_PRECLAND)) {
+uint8_t get_battery_failsafe_action(const commander_state_s &internal_state, const uint8_t battery_warning,
+				    const low_battery_action_t param_com_low_bat_act)
+{
+	uint8_t ret = commander_state_s::MAIN_STATE_MAX;
 
-					internal_state.main_state = commander_state_s::MAIN_STATE_AUTO_RTL;
-					internal_state.timestamp = hrt_absolute_time();
-					mavlink_log_critical(mavlink_log_pub, "%s, executing RTL\t", battery_critical);
-					events::send(events::ID("commander_bat_crit_rtl"), {events::Log::Critical, events::LogInternal::Info},
-						     "Critical battery level! Executing RTL");
-				}
+	const bool already_landing = internal_state.main_state == commander_state_s::MAIN_STATE_AUTO_LAND
```

Comment:
Should this be checking nav_state instead of main_state?

Reply:
I thought not because if the main state is something else it means the vehicle can potentially switch back to that if it's available again.

I took that over to preserve the functionality added in https://github.com/PX4/PX4-Autopilot/pull/14960